File: BackEnd/DotNet/WebDigital/WebDigital/Services/Implements/MyShelfService.cs
using Microsoft.EntityFrameworkCore;
using WebDigital.dto;
using WebDigital.Exceptions;
using WebDigital.Models;
using WebDigital.Services.Interface;

namespace WebDigital.Services.Implements
{
    public class MyShelfService : IMyShelfService
    {
        private readonly BookwormfinalBackupContext _context;

        public MyShelfService(BookwormfinalBackupContext context)
        {
            _context = context;
        }

        public async Task<List<MyShelfItemDto>> GetMyShelf(int userId)
        {
            var shelfEntries = await _context.MyShelves
                .Include(s => s.Product)
                    .ThenInclude(p => p.PdfBooks)
                .Where(s => s.UserId == userId)
                .ToListAsync();

            var now = DateTime.UtcNow;

            return shelfEntries.Select(s => new MyShelfItemDto
            {
                ShelfId = s.ShelfId,
                ProductId = s.ProductId,
                ProductName = s.Product.ProductName,
                ProductImage = s.Product.ProductImage,
                ProductExpiryDate = s.ProductExpiryDate,
                IsExpired = s.ProductExpiryDate.HasValue && s.ProductExpiryDate.Value < now,
                HasPdfAvailable = s.Product.PdfBooks.Any()
            }).ToList();
        }

        public async Task<PdfBook> GetBookForReading(int userId, int productId)
        {
            // Step 1: does this user actually have a shelf entry for this product at all?
            var shelfEntry = await _context.MyShelves
                .FirstOrDefaultAsync(s => s.UserId == userId && s.ProductId == productId);

            if (shelfEntry == null)
                throw new NotFoundException("You do not own or rent this book.");

            // Step 2: if it was a rental, has it expired?
            if (shelfEntry.ProductExpiryDate.HasValue && shelfEntry.ProductExpiryDate.Value < DateTime.UtcNow)
                throw new BadRequestException("Your access to this book has expired.");

            // Step 3: does the actual PDF file even exist for this product?
            var pdfBook = await _context.PdfBooks
                .FirstOrDefaultAsync(p => p.ProductId == productId);

            if (pdfBook == null || pdfBook.PdfData == null)
                throw new NotFoundException("No readable file is available for this book.");

            return pdfBook;
        }
    }
}

File: BackEnd/Java/transaction-service/src/main/java/com/digitalshelf/transactionservice/service/MyShelfService.java
package com.digitalshelf.transactionservice.service;

import com.digitalshelf.transactionservice.client.ProductClient;
import com.digitalshelf.transactionservice.dto.MyShelfItemDto;
import com.digitalshelf.transactionservice.entity.MyShelf;
import com.digitalshelf.transactionservice.exception.BadRequestException;
import com.digitalshelf.transactionservice.exception.NotFoundException;
import com.digitalshelf.transactionservice.repository.MyShelfRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class MyShelfService {

    private final MyShelfRepository myShelfRepository;
    private final ProductClient productClient;
    private final RestTemplate loadBalancedRestTemplate; // used just for streaming the PDF bytes (see config below)

    public List<MyShelfItemDto> getMyShelf(Integer userId) {
        return myShelfRepository.findByUserId(userId).stream()
                .map(shelf -> {
                    var product = productClient.getProductSummary(shelf.getProductId());
                    boolean expired = shelf.getProductExpiryDate() != null
                            && shelf.getProductExpiryDate().isBefore(LocalDateTime.now());

                    return MyShelfItemDto.builder()
                            .shelfId(shelf.getShelfId())
                            .productId(shelf.getProductId())
                            .productName(product.getProductName())
                            .productImage(product.getProductImage())
                            .productExpiryDate(shelf.getProductExpiryDate())
                            .isExpired(expired)
                            .hasPdfAvailable(true) // simplified for now -- see note below
                            .build();
                })
                .toList();
    }

    public byte[] getBookForReading(Integer userId, Integer productId) {
        MyShelf shelfEntry = myShelfRepository.findByUserIdAndProductId(userId, productId)
                .orElseThrow(() -> new NotFoundException("You do not own or rent this book."));

        if (shelfEntry.getProductExpiryDate() != null
                && shelfEntry.getProductExpiryDate().isBefore(LocalDateTime.now())) {
            throw new BadRequestException("Your access to this book has expired.");
        }

        // Streams the PDF bytes from product-service (which owns pdf_book) via a
        // direct load-balanced call through Eureka -- Feign isn't used here since
        // Feign's default setup handles JSON, not raw binary streams, cleanly.
        try {
            return loadBalancedRestTemplate.getForObject(
                    "http://product-service/api/products/" + productId + "/pdf/raw", byte[].class);
        } catch (Exception e) {
            throw new NotFoundException("No readable file is available for this book.");
        }
    }
}